File: wrapper/JSciSDK/JSciSDK/src/main/java/com/nuclearinstruments/jscisdk/SciSDK.java
package com.nuclearinstruments.jscisdk;

import com.sun.jna.*;
import com.sun.jna.ptr.*;

public class SciSDK {

    // pointer to the scisdk class object
    private Pointer scisdk_handle;

    public SciSDK() {
        scisdk_handle = SciSDKLibrary.INSTANCE.SCISDK_InitLib();
    }

    public int FreeLib() {
        return SciSDKLibrary.INSTANCE.SCISDK_FreeLib(scisdk_handle);
    }

    // Method used to add new device
    public int AddNewDevice(String device_path, String model, String json_file_path, String name) {
        return SciSDKLibrary.INSTANCE.SCISDK_AddNewDevice(device_path, model, json_file_path, name, scisdk_handle);
    }

    // Method used to detach device
    public int DetachDevice(String board_name) {
        return SciSDKLibrary.INSTANCE.SCISDK_DetachDevice(board_name, scisdk_handle);
    }

    // Method used to set register value
    public int SetRegister(String path, int value) {
        return SciSDKLibrary.INSTANCE.SCISDK_SetRegister(path, value, scisdk_handle);
    }

    // Method used to retrieve register value
    public int GetRegister(String path, Ref<Integer> value) {
        IntByReference return_value = new IntByReference();
        int res = SciSDKLibrary.INSTANCE.SCISDK_GetRegister(path, return_value, scisdk_handle);
        value.set(return_value.getValue());
        return res;
    }

    // Method used to get error description
    public int s_error(int error, Ref<String> value) {
        //Pointer value_pointer = new Memory(1);
        String[] value_pointer = new String[1];
        int res = SciSDKLibrary.INSTANCE.SCISDK_s_error(error, value_pointer, scisdk_handle);
        String error_description = value_pointer[0];
        value.set(error_description);
        return res;
    }

    // method used to set parameter integer value
    public int SetParameterInteger(String path, int value) {
        return SciSDKLibrary.INSTANCE.SCISDK_SetParameterInteger(path, value, scisdk_handle);
    }

    // method used to set parameter double value
    public int SetParameterDouble(String path, double value) {
        return SciSDKLibrary.INSTANCE.SCISDK_SetParameterDouble(path, value, scisdk_handle);
    }

    // method used to set parameter string value
    public int SetParameterString(String path, String value) {
        return SciSDKLibrary.INSTANCE.SCISDK_SetParameterString(path, value, scisdk_handle);
    }

    // method used to get parameter integer value
    public int GetParameterInteger(String path, Ref<Integer> value) {
        IntByReference return_value = new IntByReference();
        int res = SciSDKLibrary.INSTANCE.SCISDK_GetParameterInteger(path, return_value, scisdk_handle);
        value.set(return_value.getValue());
        return res;
    }

    // method used to get parameter double value
    public int GetParameterDouble(String path, Ref<Double> value) {
        DoubleByReference return_value = new DoubleByReference();
        int res = SciSDKLibrary.INSTANCE.SCISDK_GetParameterDouble(path, return_value, scisdk_handle);
        value.set(return_value.getValue());
        return res;
    }

    // method used to get parameter string value
    public int GetParameterString(String path, Ref<String> value) {
        String[] return_value = new String[1];
        int res = SciSDKLibrary.INSTANCE.SCISDK_GetParameterString(path, return_value, scisdk_handle);
        value.set(return_value[0]);
        return res;
    }

    // method used to execute command
    public int ExecuteCommand(String path, String value) {
        return SciSDKLibrary.INSTANCE.SCISDK_ExecuteCommand(path, value, scisdk_handle);
    }

    // allocate buffer
    public int AllocateBuffer(String path, Ref buffer) {
        // SPECTRUM DECODED BUFFER
        if (buffer.get().getClass().equals(SpectrumDecodedBuffer.class)) {
            SpectrumDecodedBuffer[] s = new SpectrumDecodedBuffer[1];
            s[0] = new SpectrumDecodedBuffer();
            int res = SciSDKLibrary.INSTANCE.SCISDK_AllocateBuffer(path, 1, s, scisdk_handle);
            buffer.set(s[0]);
            return res;
        } // OSCILLOSCOPE DECODED BUFFER
        else if (buffer.get().getClass().equals(OscilloscopeDecodedBuffer.class)) {
            OscilloscopeDecodedBuffer[] s = new OscilloscopeDecodedBuffer[1];
            s[0] = new OscilloscopeDecodedBuffer();
            int res = SciSDKLibrary.INSTANCE.SCISDK_AllocateBuffer(path, 1, s, scisdk_handle);
            buffer.set(s[0]);
            return res;
        }
        return -1;
    }

    // allocate buffer with specified size
    public int AllocateBuffer(String path, Ref buffer, int size) {
        // SPECTRUM DECODED BUFFER
        if (buffer.get().getClass().equals(SpectrumDecodedBuffer.class)) {
            SpectrumDecodedBuffer[] s = new SpectrumDecodedBuffer[1];
            s[0] = new SpectrumDecodedBuffer();
            int res = SciSDKLibrary.INSTANCE.SCISDK_AllocateBufferSize(path, 1, s, scisdk_handle, size);
            buffer.set(s[0]);
            return res;
        } // OSCILLOSCOPE DECODED BUFFER
        else if (buffer.get().getClass().equals(OscilloscopeDecodedBuffer.class)) {
            OscilloscopeDecodedBuffer[] s = new OscilloscopeDecodedBuffer[1];
            s[0] = new OscilloscopeDecodedBuffer();
            int res = SciSDKLibrary.INSTANCE.SCISDK_AllocateBufferSize(path, 1, s, scisdk_handle, size);
            buffer.set(s[0]);
            return res;
        }
        return -1;
    }

    // read data
    public int ReadData(String path, Ref buffer) {
        // SPECTRUM DECODED BUFFER
        if (buffer.get().getClass().equals(SpectrumDecodedBuffer.class)) {
            SpectrumDecodedBuffer buf = (SpectrumDecodedBuffer) buffer.get();
            return SciSDKLibrary.INSTANCE.SCISDK_ReadData(path, buf, scisdk_handle);
        } // OSCILLOSCOPE DECODED BUFFER
        else if (buffer.get().getClass().equals(OscilloscopeDecodedBuffer.class)) {
            OscilloscopeDecodedBuffer buf = (OscilloscopeDecodedBuffer) buffer.get();
            return SciSDKLibrary.INSTANCE.SCISDK_ReadData(path, buf, scisdk_handle);
        }
        return -1;
    }

    // make free buffer's memory
    public int FreeBuffer(String path, Ref buffer) {
        // SPECTRUM DECODED BUFFER
        if (buffer.get().getClass().equals(SpectrumDecodedBuffer.class)) {
            SpectrumDecodedBuffer[] s = new SpectrumDecodedBuffer[1];
            s[0] = (SpectrumDecodedBuffer) buffer.get();
            return SciSDKLibrary.INSTANCE.SCISDK_FreeBuffer(path, 1, s, scisdk_handle);
        } // OSCILLOSCOPE DECODED BUFFER
        else if (buffer.get().getClass().equals(OscilloscopeDecodedBuffer.class)) {
            OscilloscopeDecodedBuffer[] s = new OscilloscopeDecodedBuffer[1];
            s[0] = (OscilloscopeDecodedBuffer) buffer.get();
            return SciSDKLibrary.INSTANCE.SCISDK_FreeBuffer(path, 1, s, scisdk_handle);
        }

        return -1;
    }

    // get component list
    public int GetComponentList(String name, String type, Ref<String> ret, boolean return_json) {
        String ret_string[] = new String[1];
        int res = SciSDKLibrary.INSTANCE.SCISDK_GetComponentList(name, type, ret_string, return_json, scisdk_handle);
        ret.set(ret_string[0]);
        return res;
    }

    // get all parameters
    public int GetAllParameters(String path, Ref<String> ret) {
        String ret_string[] = new String[1];
        int res = SciSDKLibrary.INSTANCE.SCISDK_GetAllParameters(path, ret_string, scisdk_handle);
        ret.set(ret_string[0]);
        return res;
    }

    // get parameter description
    public int GetParameterDescription(String path, Ref<String> ret) {
        String ret_string[] = new String[1];
        int res = SciSDKLibrary.INSTANCE.SCISDK_GetParameterDescription(path, ret_string, scisdk_handle);
        ret.set(ret_string[0]);
        return res;
    }

    // get parameter list of values
    public int GetParameterListOfValues(String path, Ref<String> ret) {
        String ret_string[] = new String[1];
        int res = SciSDKLibrary.INSTANCE.SCISDK_GetParameterListOfValues(path, ret_string, scisdk_handle);
        ret.set(ret_string[0]);
        return res;
    }

    // get parameter minimum value
    public int GetParameterMinimumValue(String path, Ref<Double> ret) {
        double[] ret_value = new double[1];
        int res = SciSDKLibrary.INSTANCE.SCISDK_GetParameterMinimumValue(path, ret_value, scisdk_handle);
        ret.set(ret_value[0]);
        return res;
    }

    // get parameter maximum value
    public int GetParameterMaximumValue(String path, Ref<Double> ret) {
        double[] ret_value = new double[1];
        int res = SciSDKLibrary.INSTANCE.SCISDK_GetParameterMaximumValue(path, ret_value, scisdk_handle);
        System.out.println(ret_value[0]);
        ret.set(ret_value[0]);
        return res;
    }

    // get parameter properties
    public int GetParameterProperties(String path, Ref<String> ret) {
        String ret_string[] = new String[1];
        int res = SciSDKLibrary.INSTANCE.SCISDK_GetParametersProperties(path, ret_string, scisdk_handle);
        ret.set(ret_string[0]);
        return res;
    }

    public int FreeString(String _str) {
        return SciSDKLibrary.INSTANCE.SCISDK_free_string(_str);
    }

}

File: wrapper/Typescript/src/SciSDK.ts
import { Pointer, alloc, types, ref, refType } from 'ref-napi';
import { SciSDKInterface } from './SciSDKInterface';
import { SciSDKCustomPacketDecodedBuffer, SciSDKDigitizerDecodedBuffer, SciSDKDigitizerRawBuffer, SciSDKFFTDecodedBuffer, SciSDKFFTRawBuffer, SciSDKFFTStatus, SciSDKFrameDecodedBuffer, SciSDKFrameRawBuffer, SciSDKListRawBuffer, SciSDKOscilloscopeDecodedBuffer, SciSDKOscilloscopeDualDecodedBuffer, SciSDKOscilloscopeDualRawBuffer, SciSDKOscilloscopeRawBuffer, SciSDKOscilloscopeStatus, SciSDKRateMeterRawBuffer, SciSDKSpectrumDecodedBuffer, SciSDKSpectrumStatus } from './SciSDKDefines';


export class SciSDK {

    private handle: any = null;

    constructor() {
        // init library
        this.handle = SciSDKInterface.SCISDK_InitLib();
    }

    FreeLibrary(): number {
        return SciSDKInterface.SCISDK_FreeLib(this.handle);
    }

    AddNewDevice(devicePath: string, deviceModel: string, JSONFwFilePath: string, name: string): number {
        return SciSDKInterface.SCISDK_AddNewDevice(devicePath, deviceModel, JSONFwFilePath, name, this.handle);
    }

    DetachDevice(name: string): number {
        return SciSDKInterface.SCISDK_DetachDevice(name, this.handle);
    }

    SetParameterString(path: string, value: string): number {
        return SciSDKInterface.SCISDK_SetParameterString(path, value, this.handle);
    }

    SetParameterInteger(path: string, value: number): number {
        if (!Number.isInteger(value)) {
            throw new Error("Value is not an integer number");
        }
        return SciSDKInterface.SCISDK_SetParameterInteger(path, value, this.handle);
    }

    SetParameterUInteger(path: string, value: number): number {
        if (!Number.isInteger(value) || value < 0) {
            throw new Error("Value is not an unsigned integer number");
        }
        return SciSDKInterface.SCISDK_SetParameterUInteger(path, value, this.handle);
    }

    SetParameterDouble(path: string, value: number): number {
        return SciSDKInterface.SCISDK_SetParameterDouble(path, value, this.handle);
    }

    GetParameterString(path: string): NumberStringTuple {
        var ret_ptr: Pointer<any> = <Pointer<any>>alloc(types.CString);
        let res = SciSDKInterface.SCISDK_GetParameterString(path, ret_ptr, this.handle);
        return [res, ret_ptr.readPointer().readCString()];
    }

    GetParameterInteger(path: string): NumberNumberTuple {
        let ret_ptr: Pointer<number> = alloc(types.uint32);
        let res = SciSDKInterface.SCISDK_GetParameterInteger(path, ret_ptr, this.handle);
        return [res, ret_ptr.readUInt32LE()];
    }

    GetParameterUInteger(path: string): NumberNumberTuple {
        let ret_ptr: Pointer<number> = alloc(types.uint32);
        let res = SciSDKInterface.SCISDK_GetParameterUInteger(path, ret_ptr, this.handle);
        return [res, ret_ptr.readUInt32LE()];
    }

    GetParameterDouble(path: string): NumberNumberTuple {
        let ret_ptr: Pointer<number> = alloc(types.uint32);
        let res = SciSDKInterface.SCISDK_GetParameterDouble(path, ret_ptr, this.handle);
        return [res, ret_ptr.readUInt32LE()];
    }

    SetRegister(path: string, value: number): number {
        if (!Number.isInteger(value) || value < 0) {
            throw new Error("Value is not an unsigned integer number");
        }

        return SciSDKInterface.SCISDK_SetRegister(path, value, this.handle);
    }

    GetRegister(path: string): NumberNumberTuple {
        let ret_ptr: Pointer<number> = alloc(types.uint32);
        let res = SciSDKInterface.SCISDK_GetRegister(path, ret_ptr, this.handle);
        return [res, ret_ptr.readUInt32LE()];
    }

    AllocateBuffer(path: string, buffer: any): number {
        let decoded = 0;
        if (buffer instanceof SciSDKOscilloscopeDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKOscilloscopeDecodedBuffer.cpointer_class, new SciSDKOscilloscopeDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKOscilloscopeRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKOscilloscopeRawBuffer.cpointer_class, new SciSDKOscilloscopeRawBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKOscilloscopeDualDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKOscilloscopeDualDecodedBuffer.cpointer_class, new SciSDKOscilloscopeDualDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKOscilloscopeDualRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKOscilloscopeDualRawBuffer.cpointer_class, new SciSDKOscilloscopeDualRawBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKDigitizerDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKDigitizerDecodedBuffer.cpointer_class, new SciSDKDigitizerDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKDigitizerRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKDigitizerRawBuffer.cpointer_class, new SciSDKDigitizerRawBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKCustomPacketDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKCustomPacketDecodedBuffer.cpointer_class, new SciSDKCustomPacketDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKListRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKListRawBuffer.cpointer_class, new SciSDKListRawBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKRateMeterRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKRateMeterRawBuffer.cpointer_class, new SciSDKRateMeterRawBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKSpectrumDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKSpectrumDecodedBuffer.cpointer_class, new SciSDKSpectrumDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKFFTDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKFFTDecodedBuffer.cpointer_class, new SciSDKFFTDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKFFTRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKFFTRawBuffer.cpointer_class, new SciSDKFFTRawBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKFrameDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKFrameDecodedBuffer.cpointer_class, new SciSDKFrameDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKFrameRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKFrameRawBuffer.cpointer_class, new SciSDKFrameRawBuffer.cpointer_class).ref();
        }
        else {
            throw new Error("The type of the buffer is invalid");
        }

        let res = SciSDKInterface.SCISDK_AllocateBuffer(path, decoded, buffer.cpointer, this.handle);
        buffer.cpointer = buffer.cpointer.deref().deref()
        buffer.LoadData();
        return res;
    }

    AllocateBufferSize(path: string, buffer: any, size: number): number {
        let decoded = 0;
        if (buffer instanceof SciSDKOscilloscopeDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKOscilloscopeDecodedBuffer.cpointer_class, new SciSDKOscilloscopeDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKOscilloscopeRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKOscilloscopeRawBuffer.cpointer_class, new SciSDKOscilloscopeRawBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKOscilloscopeDualDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKOscilloscopeDualDecodedBuffer.cpointer_class, new SciSDKOscilloscopeDualDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKOscilloscopeDualRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKOscilloscopeDualRawBuffer.cpointer_class, new SciSDKOscilloscopeDualRawBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKDigitizerDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKDigitizerDecodedBuffer.cpointer_class, new SciSDKDigitizerDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKDigitizerRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKDigitizerRawBuffer.cpointer_class, new SciSDKDigitizerRawBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKListRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKListRawBuffer.cpointer_class, new SciSDKListRawBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKCustomPacketDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKCustomPacketDecodedBuffer.cpointer_class, new SciSDKCustomPacketDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKRateMeterRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKRateMeterRawBuffer.cpointer_class, new SciSDKRateMeterRawBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKSpectrumDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKSpectrumDecodedBuffer.cpointer_class, new SciSDKSpectrumDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKFFTDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKFFTDecodedBuffer.cpointer_class, new SciSDKFFTDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKFFTRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKFFTRawBuffer.cpointer_class, new SciSDKFFTRawBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKFrameDecodedBuffer) {
            decoded = 1;
            buffer.cpointer = alloc(SciSDKFrameDecodedBuffer.cpointer_class, new SciSDKFrameDecodedBuffer.cpointer_class).ref();
        }
        else if (buffer instanceof SciSDKFrameRawBuffer) {
            decoded = 0;
            buffer.cpointer = alloc(SciSDKFrameRawBuffer.cpointer_class, new SciSDKFrameRawBuffer.cpointer_class).ref();
        }
        else {
            throw new Error("The type of the buffer is invalid");
        }

        let res = SciSDKInterface.SCISDK_AllocateBufferSize(path, decoded, buffer.cpointer, this.handle, size);
        buffer.cpointer = buffer.cpointer.deref().deref()
        buffer.LoadData();
        return res;
    }

    ReadData(path: string, buffer: any) {
        let res = -1;
        if (buffer instanceof SciSDKOscilloscopeDecodedBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKOscilloscopeDecodedBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else if (buffer instanceof SciSDKOscilloscopeRawBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKOscilloscopeRawBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else if (buffer instanceof SciSDKOscilloscopeDualDecodedBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKOscilloscopeDualDecodedBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else if (buffer instanceof SciSDKOscilloscopeDualRawBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKOscilloscopeDualRawBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else if (buffer instanceof SciSDKDigitizerDecodedBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKDigitizerDecodedBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else if (buffer instanceof SciSDKDigitizerRawBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKDigitizerRawBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else if (buffer instanceof SciSDKListRawBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKListRawBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else if (buffer instanceof SciSDKCustomPacketDecodedBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKCustomPacketDecodedBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else if (buffer instanceof SciSDKRateMeterRawBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKRateMeterRawBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else if (buffer instanceof SciSDKSpectrumDecodedBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKSpectrumDecodedBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else if (buffer instanceof SciSDKFFTDecodedBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKFFTDecodedBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else if (buffer instanceof SciSDKFFTRawBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKFFTRawBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else if (buffer instanceof SciSDKFrameDecodedBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKFrameDecodedBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else if (buffer instanceof SciSDKFrameRawBuffer) {
            res = SciSDKInterface.SCISDK_ReadData(path, (<Pointer<typeof SciSDKFrameRawBuffer.cpointer_class>>buffer.cpointer).ref(), this.handle);
        }
        else {
            throw new Error("The type of the buffer is invalid");
        }

        buffer.LoadData();
        return res;
    }

    FreeBuffer(path: string, buffer: any): number {
        let decoded = 0;
        if (buffer instanceof SciSDKOscilloscopeDecodedBuffer) {
            decoded = 1;
        }
        else if (buffer instanceof SciSDKOscilloscopeRawBuffer) {
            decoded = 0;
        }
        else if (buffer instanceof SciSDKOscilloscopeDualDecodedBuffer) {
            decoded = 1;
        }
        else if (buffer instanceof SciSDKOscilloscopeDualRawBuffer) {
            decoded = 0;
        }
        else if (buffer instanceof SciSDKDigitizerDecodedBuffer) {
            decoded = 1;
        }
        else if (buffer instanceof SciSDKDigitizerRawBuffer) {
            decoded = 0;
        }
        else if (buffer instanceof SciSDKListRawBuffer) {
            decoded = 0;
        }
        else if (buffer instanceof SciSDKCustomPacketDecodedBuffer) {
            decoded = 1;
        }
        else if (buffer instanceof SciSDKRateMeterRawBuffer) {
            decoded = 0;
        }
        else if (buffer instanceof SciSDKSpectrumDecodedBuffer) {
            decoded = 1;
        }
        else if (buffer instanceof SciSDKFFTDecodedBuffer) {
            decoded = 1;
        }
        else if (buffer instanceof SciSDKFFTRawBuffer) {
            decoded = 0;
        }
        else if (buffer instanceof SciSDKFrameDecodedBuffer) {
            decoded = 1;
        }
        else if (buffer instanceof SciSDKFrameRawBuffer) {
            decoded = 0;
        }
        else {
            throw new Error("The type of the buffer is invalid");
        }
        let res = SciSDKInterface.SCISDK_FreeBuffer(path, decoded, buffer.cpointer.ref(), this.handle);
        buffer.cpointer = null;
        return res;
    }

    ReadStatus(path: string, buffer: any): number {
        
        // if (buffer instanceof SciSDKOscilloscopeStatus) {
        //     buffer.cpointer 
        // }
        // else if (buffer instanceof SciSDKFFTStatus) {

        // }
        // else if (buffer instanceof SciSDKSpectrumStatus) {

        // }
        // else {
        //     throw new Error("The type of the buffer is invalid");
        // }
        return 0;
    }

    DecodeData(path: string, buffer_in: any, buffer_out: any): number {
        // TODO: check if buffers are valid type
        return SciSDKInterface.SCISDK_DecodeData(path, buffer_in.cpointer, buffer_out.cpointer, this.handle);
    }

    ExecuteCommand(path: string, value: string): number {
        return SciSDKInterface.SCISDK_ExecuteCommand(path, value, this.handle);
    }

    s_error(ni_error: number): NumberStringTuple {
        var ret_ptr: Pointer<any> = <Pointer<any>>alloc(types.CString);
        let res = SciSDKInterface.SCISDK_s_error(ni_error, ret_ptr, this.handle);
        return [res, ret_ptr.readPointer().readCString()];
    }

    GetComponentList(board: string, type: string, return_json: boolean): NumberStringTuple {
        var ret_ptr: Pointer<any> = <Pointer<any>>alloc(types.CString);
        let res = SciSDKInterface.SCISDK_GetComponentList(board, type, ret_ptr, return_json, this.handle);
        return [res, ret_ptr.readPointer().readCString()];
    }

    GetAllParameters(path: string): NumberStringTuple {
        var ret_ptr: Pointer<any> = <Pointer<any>>alloc(types.CString);
        let res = SciSDKInterface.SCISDK_GetAllParameters(path, ret_ptr, this.handle);
        return [res, ret_ptr.readPointer().readCString()];
    }

    GetParameterDescription(path: string): NumberStringTuple {
        var ret_ptr: Pointer<any> = <Pointer<any>>alloc(types.CString);
        let res = SciSDKInterface.SCISDK_GetParameterDescription(path, ret_ptr, this.handle);
        return [res, ret_ptr.readPointer().readCString()];
    }

    GetParameterListOfValues(path: string): NumberStringTuple {
        var ret_ptr: Pointer<any> = <Pointer<any>>alloc(types.CString);
        let res = SciSDKInterface.SCISDK_GetParameterListOfValues(path, ret_ptr, this.handle);
        return [res, ret_ptr.readPointer().readCString()];
    }

    GetParameterMinimumValue(path: string): NumberNumberTuple {
        var ret_ptr: Pointer<number> = <Pointer<number>>alloc(types.double);
        let res = SciSDKInterface.SCISDK_GetParameterMinimumValue(path, ret_ptr, this.handle);
        return [res, ret_ptr.readDoubleBE()];
    }

    GetParametersProperties(path: string): NumberStringTuple {
        var ret_ptr: Pointer<any> = <Pointer<any>>alloc(types.CString);
        let res = SciSDKInterface.SCISDK_GetParametersProperties(path, ret_ptr, this.handle);
        return [res, ret_ptr.readPointer().readCString()];
    }

    free_string(str: string): number {
        // return SciSDKInterface.SCISDK_free_string(str);
        return 0;
    }

    GetAttachedDevicesList(): NumberStringTuple {
        var ret_ptr: Pointer<any> = <Pointer<any>>alloc(types.CString);
        let res = SciSDKInterface.SCISDK_GetAttachedDevicesList(ret_ptr, this.handle);
        return [res, ret_ptr.readPointer().readCString()];
    }

    GetLibraryVersion(): NumberStringTuple {
        var ret_ptr: Pointer<any> = <Pointer<any>>alloc(types.CString);
        let res = SciSDKInterface.SCISDK_GetLibraryVersion(ret_ptr, this.handle);
        return [res, ret_ptr.readPointer().readCString()];
    }

}

export interface NumberStringTuple extends Array<string | number> { 0: number; 1: string }
export interface NumberNumberTuple extends Array<number> { 0: number; 1: number }